### PR TITLE
Add bit-count and bit-reverse AArch64 instructions (#58)

### PR DESCRIPTION
**Pull request by pmatos**
## Summary

- New AArch64 IR variants: CLZ, CLS, RBIT, REV, REV32, REV16 (all `{rd, rn}`).
- Concrete semantics (`leading_zeros` / `reverse_bits` / `swap_bytes`) and Z3 SMT lowering (extract+concat permutations for REV/REV32/REV16/RBIT; 64-deep ITE chain for CLZ; CLS reduces to CLZ via the `x ^ (x ASR 63)` sign-fold).
- Wired through the cost model, dynasm encoder, GAS parser, enumerative and stochastic candidate generation, and the MCMC peer-mutation cluster.
- Always encodable (register-only), latency 1 each.

Fixes #58

## Test plan

- [x] `cargo fmt -- --check` — clean
- [x] `cargo build` — clean
- [x] `cargo test` — full suite passes (`./ci_check.sh`)
- [x] Concrete tests cross-check REV against `u64::swap_bytes`, RBIT against `u64::reverse_bits`, CLZ against `u64::leading_zeros`, and add CLS spec-vector tests
- [x] Encoder round-trip via Capstone for all 6 mnemonics
- [x] SMT involution proofs for REV / REV32 / REV16 / RBIT
- [x] SMT floor-log2 acceptance test: `CLZ x0, x1; MOV x2, #63; SUB x0, x2, x0` produces the highest-set-bit position for nonzero `x1` (the issue's `modulo zero-input edge case` example)

### CODE REVIEW

File: src/ir/instructions.rs [hunk 1]
```diff
@@ -198,6 +198,32 @@ pub enum Instruction {
         rn: Register,
         shift: Operand,
     },
+
+    // Single-source bit-manipulation (always register-only, {rd, rn})
+    Clz {
+        rd: Register,
+        rn: Register,
+    },
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/ir/instructions.rs:226]: **Out of scope for this PR** — confirmed real, but pre-existing for every IR instruction except `mov`/`add`/`nop`. `convert_to_ir` (src/main.rs:775-807) silently warn-and-skips ~30 already-supported mnemonics today (SUB, MUL, SDIV, CSEL, MVN, NEG, NEGS, MOVN, BIC/BICS/ORN/EON, ADDS/SUBS/ANDS, CSET/CSETM, ROR, LSL/LSR/ASR, …). The six new bit-manip ops are no different from any of those; the correct fix is rewriting the Capstone→IR converter (most likely delegating to `parser::parse_line` on the disassembled text) and adding regression coverage for the full instruction set, which is orthogonal to this PR's "add 6 IR variants" goal. Patching only the bit-manip subset here would deepen the existing inconsistency.

Tracked as #110 (https://github.com/pmatos/s11/issues/110), which bundles the broader fix together with @claude's related "ELF pipeline not directly tested" follow-up suggestion (same root cause).

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.x). If you disagree, please reply and we'll re-evaluate._

File: src/semantics/smt.rs [hunk 2]
```diff
@@ -20,6 +20,46 @@
...
+/// Reverse the byte order of a 64-bit BV by concatenating its 8 byte slices
+/// with byte 0 placed in the most-significant position.
+fn bv_swap_bytes_64(value: &BV) -> BV {
+    // Place byte 0 (originally at bits [7:0]) at the new top, byte 7
+    // (originally at bits [63:56]) at the new bottom.
+    let mut result = value.extract(7, 0);
+    for i in 1..8u32 {
+        let lo = i * 8;
+        let hi = lo + 7;
+        result = result.concat(&value.extract(hi, lo));
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs:32]: **Deferred (diminishing returns)** — pure-lint nit; pre-existing equivalents on lines 494/521/548/587 of the same file are not flagged because they aren't in this diff. Fixing only the new lines would leave the file stylistically inconsistent (half `&x`, half `x` for the same Z3 API). Right shape is a whole-file (or repo-wide) clippy sweep in its own PR, not a per-PR patch on just the diff.

Tracked as #111 (https://github.com/pmatos/s11/issues/111), which lists all 9 unique sites in `src/semantics/smt.rs` plus the 4 pre-existing equivalents to include in the sweep.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.x). If you disagree, please reply and we'll re-evaluate._

[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs:32]: **Deferred (diminishing returns)** — pure-lint nit; pre-existing equivalents on lines 494/521/548/587 of the same file are not flagged because they aren't in this diff. Fixing only the new lines would leave the file stylistically inconsistent (half `&x`, half `x` for the same Z3 API). Right shape is a whole-file (or repo-wide) clippy sweep in its own PR, not a per-PR patch on just the diff.

Tracked as #111 (https://github.com/pmatos/s11/issues/111), which lists all 9 unique sites in `src/semantics/smt.rs` plus the 4 pre-existing equivalents to include in the sweep.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.x). If you disagree, please reply and we'll re-evaluate._

File: src/semantics/smt.rs [hunk 3]
```diff
@@ -20,6 +20,46 @@
...
+    }
+    result
+}
+
+/// Reverse the bit order of a 64-bit BV via 64 single-bit extracts.
+fn bv_reverse_bits_64(value: &BV) -> BV {
+    // Bit 0 of `value` becomes the new MSB; bit 63 becomes the new LSB.
+    let mut result = value.extract(0, 0);
+    for i in 1..64u32 {
+        result = result.concat(&value.extract(i, i));
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs:42]: **Deferred (diminishing returns)** — pure-lint nit; pre-existing equivalents on lines 494/521/548/587 of the same file are not flagged because they aren't in this diff. Fixing only the new lines would leave the file stylistically inconsistent (half `&x`, half `x` for the same Z3 API). Right shape is a whole-file (or repo-wide) clippy sweep in its own PR, not a per-PR patch on just the diff.

Tracked as #111 (https://github.com/pmatos/s11/issues/111), which lists all 9 unique sites in `src/semantics/smt.rs` plus the 4 pre-existing equivalents to include in the sweep.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.x). If you disagree, please reply and we'll re-evaluate._

[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs:42]: **Deferred (diminishing returns)** — pure-lint nit; pre-existing equivalents on lines 494/521/548/587 of the same file are not flagged because they aren't in this diff. Fixing only the new lines would leave the file stylistically inconsistent (half `&x`, half `x` for the same Z3 API). Right shape is a whole-file (or repo-wide) clippy sweep in its own PR, not a per-PR patch on just the diff.

Tracked as #111 (https://github.com/pmatos/s11/issues/111), which lists all 9 unique sites in `src/semantics/smt.rs` plus the 4 pre-existing equivalents to include in the sweep.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.x). If you disagree, please reply and we'll re-evaluate._

File: src/semantics/smt.rs [hunk 4]
```diff
@@ -308,6 +348,68 @@
...
+            state.set_register(*rd, bv_clz_64(&value));
+        }
+        // CLS: count leading sign-bit replicas (excluding the sign bit).
+        // Fold the sign bit out via `x XOR (x ASR 63)` so the answer reduces
+        // to `clz(folded) - 1`. For all-sign inputs (0 or -1) folded is zero,
+        // clz is 64, and the result clamps to 63 via bvsub (no underflow at
+        // 64-bit width).
+        Instruction::Cls { rd, rn } => {
+            let value = state.get_register(*rn).clone();
+            let asr = value.bvashr(&BV::from_u64(63, 64));
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs:383]: **Deferred (diminishing returns)** — pure-lint nit; pre-existing equivalents on lines 494/521/548/587 of the same file are not flagged because they aren't in this diff. Fixing only the new lines would leave the file stylistically inconsistent (half `&x`, half `x` for the same Z3 API). Right shape is a whole-file (or repo-wide) clippy sweep in its own PR, not a per-PR patch on just the diff.

Tracked as #111 (https://github.com/pmatos/s11/issues/111), which lists all 9 unique sites in `src/semantics/smt.rs` plus the 4 pre-existing equivalents to include in the sweep.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.x). If you disagree, please reply and we'll re-evaluate._

[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs:383]: **Deferred (diminishing returns)** — pure-lint nit; pre-existing equivalents on lines 494/521/548/587 of the same file are not flagged because they aren't in this diff. Fixing only the new lines would leave the file stylistically inconsistent (half `&x`, half `x` for the same Z3 API). Right shape is a whole-file (or repo-wide) clippy sweep in its own PR, not a per-PR patch on just the diff.

Tracked as #111 (https://github.com/pmatos/s11/issues/111), which lists all 9 unique sites in `src/semantics/smt.rs` plus the 4 pre-existing equivalents to include in the sweep.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.x). If you disagree, please reply and we'll re-evaluate._

File: src/semantics/smt.rs [hunk 5]
```diff
@@ -308,6 +348,68 @@
...
+            let folded = value.bvxor(&asr);
+            let clz = bv_clz_64(&folded);
+            let result = clz.bvsub(&BV::from_u64(1, 64));
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs:386]: **Deferred (diminishing returns)** — pure-lint nit; pre-existing equivalents on lines 494/521/548/587 of the same file are not flagged because they aren't in this diff. Fixing only the new lines would leave the file stylistically inconsistent (half `&x`, half `x` for the same Z3 API). Right shape is a whole-file (or repo-wide) clippy sweep in its own PR, not a per-PR patch on just the diff.

Tracked as #111 (https://github.com/pmatos/s11/issues/111), which lists all 9 unique sites in `src/semantics/smt.rs` plus the 4 pre-existing equivalents to include in the sweep.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.x). If you disagree, please reply and we'll re-evaluate._

[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs:386]: **Deferred (diminishing returns)** — pure-lint nit; pre-existing equivalents on lines 494/521/548/587 of the same file are not flagged because they aren't in this diff. Fixing only the new lines would leave the file stylistically inconsistent (half `&x`, half `x` for the same Z3 API). Right shape is a whole-file (or repo-wide) clippy sweep in its own PR, not a per-PR patch on just the diff.

Tracked as #111 (https://github.com/pmatos/s11/issues/111), which lists all 9 unique sites in `src/semantics/smt.rs` plus the 4 pre-existing equivalents to include in the sweep.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.x). If you disagree, please reply and we'll re-evaluate._

File: src/semantics/smt.rs [hunk 6]
```diff
@@ -308,6 +348,68 @@
...
+            let value = state.get_register(*rn).clone();
+            let lo = value.extract(31, 0);
+            let hi = value.extract(63, 32);
+            // Each half byte-reversed: concat its 4 byte slices with the
+            // original LSB byte at the new MSB position.
+            let rev_half = |h: &BV| -> BV {
+                let mut acc = h.extract(7, 0);
+                for i in 1..4u32 {
+                    let l = i * 8;
+                    acc = acc.concat(&h.extract(l + 7, l));
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs:410]: **Deferred (diminishing returns)** — pure-lint nit; pre-existing equivalents on lines 494/521/548/587 of the same file are not flagged because they aren't in this diff. Fixing only the new lines would leave the file stylistically inconsistent (half `&x`, half `x` for the same Z3 API). Right shape is a whole-file (or repo-wide) clippy sweep in its own PR, not a per-PR patch on just the diff.

Tracked as #111 (https://github.com/pmatos/s11/issues/111), which lists all 9 unique sites in `src/semantics/smt.rs` plus the 4 pre-existing equivalents to include in the sweep.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.x). If you disagree, please reply and we'll re-evaluate._

[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs:410]: **Deferred (diminishing returns)** — pure-lint nit; pre-existing equivalents on lines 494/521/548/587 of the same file are not flagged because they aren't in this diff. Fixing only the new lines would leave the file stylistically inconsistent (half `&x`, half `x` for the same Z3 API). Right shape is a whole-file (or repo-wide) clippy sweep in its own PR, not a per-PR patch on just the diff.

Tracked as #111 (https://github.com/pmatos/s11/issues/111), which lists all 9 unique sites in `src/semantics/smt.rs` plus the 4 pre-existing equivalents to include in the sweep.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.x). If you disagree, please reply and we'll re-evaluate._

File: src/semantics/smt.rs [hunk 7]
```diff
@@ -308,6 +348,68 @@
...
+                }
+                acc
+            };
+            let result = rev_half(&hi).concat(&rev_half(&lo));
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs:414]: **Deferred (diminishing returns)** — pure-lint nit; pre-existing equivalents on lines 494/521/548/587 of the same file are not flagged because they aren't in this diff. Fixing only the new lines would leave the file stylistically inconsistent (half `&x`, half `x` for the same Z3 API). Right shape is a whole-file (or repo-wide) clippy sweep in its own PR, not a per-PR patch on just the diff.

Tracked as #111 (https://github.com/pmatos/s11/issues/111), which lists all 9 unique sites in `src/semantics/smt.rs` plus the 4 pre-existing equivalents to include in the sweep.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.x). If you disagree, please reply and we'll re-evaluate._

[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs:414]: **Deferred (diminishing returns)** — pure-lint nit; pre-existing equivalents on lines 494/521/548/587 of the same file are not flagged because they aren't in this diff. Fixing only the new lines would leave the file stylistically inconsistent (half `&x`, half `x` for the same Z3 API). Right shape is a whole-file (or repo-wide) clippy sweep in its own PR, not a per-PR patch on just the diff.

Tracked as #111 (https://github.com/pmatos/s11/issues/111), which lists all 9 unique sites in `src/semantics/smt.rs` plus the 4 pre-existing equivalents to include in the sweep.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.x). If you disagree, please reply and we'll re-evaluate._

File: src/semantics/smt.rs [hunk 8]
```diff
@@ -308,6 +348,68 @@
...
+            state.set_register(*rd, result);
+        }
+        // REV16: byte-reverse within each 16-bit half (four halves).
+        Instruction::Rev16 { rd, rn } => {
+            let value = state.get_register(*rn).clone();
+            // For each of the 4 half-words, swap its high and low byte.
+            let swap_half = |start: u32| -> BV {
+                value
+                    .extract(start + 7, start)
+                    .concat(&value.extract(start + 15, start + 8))
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs:424]: **Deferred (diminishing returns)** — pure-lint nit; pre-existing equivalents on lines 494/521/548/587 of the same file are not flagged because they aren't in this diff. Fixing only the new lines would leave the file stylistically inconsistent (half `&x`, half `x` for the same Z3 API). Right shape is a whole-file (or repo-wide) clippy sweep in its own PR, not a per-PR patch on just the diff.

Tracked as #111 (https://github.com/pmatos/s11/issues/111), which lists all 9 unique sites in `src/semantics/smt.rs` plus the 4 pre-existing equivalents to include in the sweep.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.x). If you disagree, please reply and we'll re-evaluate._

[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs:424]: **Deferred (diminishing returns)** — pure-lint nit; pre-existing equivalents on lines 494/521/548/587 of the same file are not flagged because they aren't in this diff. Fixing only the new lines would leave the file stylistically inconsistent (half `&x`, half `x` for the same Z3 API). Right shape is a whole-file (or repo-wide) clippy sweep in its own PR, not a per-PR patch on just the diff.

Tracked as #111 (https://github.com/pmatos/s11/issues/111), which lists all 9 unique sites in `src/semantics/smt.rs` plus the 4 pre-existing equivalents to include in the sweep.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.x). If you disagree, please reply and we'll re-evaluate._

File: src/semantics/smt.rs [hunk 9]
```diff
@@ -489,4 +591,120 @@
...
+        let initial_x1 = initial.get_register(Register::X1).clone();
+        let seq = vec![
+            op(Register::X0, Register::X1),
+            op(Register::X0, Register::X0),
+        ];
+        let final_state = apply_sequence(initial, &seq);
+        let final_x0 = final_state.get_register(Register::X0);
+
+        let solver = Solver::new();
+        solver.assert(&final_x0.eq(&initial_x1).not());
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs:626]: **Deferred (diminishing returns)** — pure-lint nit; pre-existing equivalents on lines 494/521/548/587 of the same file are not flagged because they aren't in this diff. Fixing only the new lines would leave the file stylistically inconsistent (half `&x`, half `x` for the same Z3 API). Right shape is a whole-file (or repo-wide) clippy sweep in its own PR, not a per-PR patch on just the diff.

Tracked as #111 (https://github.com/pmatos/s11/issues/111), which lists all 9 unique sites in `src/semantics/smt.rs` plus the 4 pre-existing equivalents to include in the sweep.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.x). If you disagree, please reply and we'll re-evaluate._

File: src/semantics/smt.rs [hunk 10]
```diff
@@ -489,4 +591,120 @@
...
+            },
+            Instruction::Clz {
+                rd: Register::X0,
+                rn: Register::X1,
+            },
+        ];
+        let final_state = apply_sequence(initial, &seq);
+        let final_x0 = final_state.get_register(Register::X0);
+        let solver = Solver::new();
+        solver.assert(&final_x0.eq(&BV::from_u64(63, 64)).not());
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs:673]: **Deferred (diminishing returns)** — pure-lint nit; pre-existing equivalents on lines 494/521/548/587 of the same file are not flagged because they aren't in this diff. Fixing only the new lines would leave the file stylistically inconsistent (half `&x`, half `x` for the same Z3 API). Right shape is a whole-file (or repo-wide) clippy sweep in its own PR, not a per-PR patch on just the diff.

Tracked as #111 (https://github.com/pmatos/s11/issues/111), which lists all 9 unique sites in `src/semantics/smt.rs` plus the 4 pre-existing equivalents to include in the sweep.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.x). If you disagree, please reply and we'll re-evaluate._

[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs:673]: **Deferred (diminishing returns)** — pure-lint nit; pre-existing equivalents on lines 494/521/548/587 of the same file are not flagged because they aren't in this diff. Fixing only the new lines would leave the file stylistically inconsistent (half `&x`, half `x` for the same Z3 API). Right shape is a whole-file (or repo-wide) clippy sweep in its own PR, not a per-PR patch on just the diff.

Tracked as #111 (https://github.com/pmatos/s11/issues/111), which lists all 9 unique sites in `src/semantics/smt.rs` plus the 4 pre-existing equivalents to include in the sweep.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.x). If you disagree, please reply and we'll re-evaluate._

**Comment by pmatos** [comment]
@claude Regarding your code review summary (https://github.com/pmatos/s11/pull/109#issuecomment-4441883408) — each suggestion triaged with dual-evaluator (Claude Opus 4.7 + Codex GPT-5.x), outcomes below.

---

**1. `bv_clz_64` 64-deep ITE chain — Z3 performance risk**

**Deferred (diminishing returns)** — concern is plausible but unmeasured; no current evidence of slow solving on CLZ-heavy queries. The binary-search rewrite is a non-trivial semantic change that deserves its own PR with before/after Z3 benchmarks. The current implementation is provably correct (the floor-log2 acceptance test in this PR passes), so this is purely a performance follow-up.

Tracked as #112 (https://github.com/pmatos/s11/issues/112).

---

**2. MCMC peer-mutation table is verbose**

**No action taken** — the explicit 6-way match arms intentionally mirror the existing per-instruction clusters in the same file: MVN/NEG/NEGS (3-way), BIC/BICS/ORN/EON (mixed-arity), and the CSEL family (4-way). Collapsing only the new cluster into a helper would break the established local convention; collapsing all clusters into one shared helper is a separate refactor of pre-existing code that's out of scope for adding 6 new IR variants.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.x). If you disagree, please reply and we'll re-evaluate._

---

**3. "Opcode slots" comment vs `opcode_count()` confusing dual meaning**

Fixed in `418b75e` (https://github.com/pmatos/s11/pull/109/commits/418b75e430ec5102336dfdff773c149c4b7fda7b).

Changed: the stale "24 slots / 34 IDs" wording in the `opcode_count` doc comment at `src/isa/aarch64.rs:873-880` to reflect the new 25 slots and 40 IDs after adding the six single-source bit-manipulation ops; also notes the new slot-24 sub-multiplexer.
Validation: `cargo build` and `cargo fmt -- --check` clean.

---

**4. ELF pipeline not directly tested for new mnemonics**

**Deferred (out of scope for this PR)** — same root cause as the codex P1 inline review on `src/ir/instructions.rs:206`: the ELF→IR path uses the hardcoded 3-mnemonic `convert_to_ir` in `src/main.rs:775-807`, not `parser::parse_line`. A meaningful end-to-end test on the new mnemonics requires fixing the converter first; that fix is a broad rewrite that should land its own regression coverage for all ~30 IR-supported instructions, not just the 6 added here.

Bundled into the same tracking issue as the codex P1: #110 (https://github.com/pmatos/s11/issues/110).

---

Summary of this PR's response to your review:
| Suggestion | Outcome | Reference |
|---|---|---|
| #1 `bv_clz_64` ITE depth | DEFER | #112 |
| #2 MCMC table style | REJECT | — |
| #3 opcode_count comment | FIX | 418b75e |
| #4 ELF pipeline test | DEFER | #110 |

**Comment by pmatos** [comment]
@claude Regarding your second-pass code review (https://github.com/pmatos/s11/pull/109#issuecomment-4457916285) — each suggestion triaged with dual-evaluator (Claude Opus 4.7 + Codex GPT-5.x), outcomes below.

| # | Suggestion | Outcome | Reference |
|---|---|---|---|
| a | `saturating_sub(1)` → `- 1` in CLS concrete | FIX | `99d445a` |
| b | `bv_clz_64` doc wording | FIX | `99d445a` |
| c | CLS SMT bvsub correctness comment expansion | FIX | `99d445a` |
| d | TODO(#112) signpost on `bv_clz_64` | FIX | `99d445a` |
| e | 30 top-level slots for bit-manip ops in `generate_random` | DEFER | #115 |
| f | Consolidate duplicated `opcode_id` table | DEFER | #116 |
| g | SMT CLS/CLZ equivalence proof | DEFER | #117 |

---

**FIX (all four, commit `99d445a` — https://github.com/pmatos/s11/pull/109/commits/99d445a):**

- **(a)** Replaced `(folded.leading_zeros() as u64).saturating_sub(1)` with plain `- 1` in `src/semantics/concrete.rs` and added a one-line comment recording the bit-63-zero invariant. Both evaluators agreed the `saturating_sub` was dead-code defensiveness given the sign-fold guarantees `leading_zeros >= 1`.
- **(b)** Rewrote the `bv_clz_64` doc comment in `src/semantics/smt.rs` to say the helper returns the CLZ count derived from the highest-set-bit position, rather than "reflects the most-significant set bit" (which read as if the helper returned the bit value).
- **(c)** Extended the CLS SMT inline comment to spell out the invariant chain: bit 63 of `folded` is always 0 (positive sign cancels its top bit; negative sign inverts it to 0), so `bv_clz_64(folded) ∈ [1, 64]` and the `bvsub` result lands in `[0, 63]` — replacing the "clamps via bvsub" wording you correctly flagged as misleading.
- **(d)** Added a `TODO(#112)` signpost on `bv_clz_64` linking the deferred ITE-depth follow-up to the code location.

Validation: `./ci_check.sh` clean; concrete CLS spec-vector tests and SMT CLZ tests still pass.

---

**DEFER:**

- **(e)** MCMC slot rebalancing → **#115**. The bundling matches the existing ANDS/CSET/CSETM/ROR (4-way slot 23) and singleton MVN/NEG/NEGS conventions — there is no project rule against multiplexed slots. Local evaluator REJECTed; cross-harness DEFERRed pending a stochastic-convergence benchmark. Disagreement → DEFER per the loop rules.
- **(f)** `opcode_id` duplication → **#116**. Pre-existing tech debt, predates #109. Consolidation touches an unrelated consumer; cleaner as its own PR. Local DEFERRed; cross-harness FIXed; disagreement → DEFER.
- **(g)** SMT CLS/CLZ equivalence proof → **#117**. Local DEFERRed (somewhat tautological since the lowering uses the same identity); cross-harness FIXed; disagreement → DEFER.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.x). If you disagree on any item, please reply and we'll re-evaluate._

File: src/ir/instructions.rs [hunk 11]
```diff
@@ -360,6 +392,14 @@ impl Instruction {
...
                 Operand::Immediate(amt) => *amt >= 0 && *amt <= 63,
             },
+
+            // Single-source bit-manipulation: always encodable (register-only).
+            Instruction::Clz { .. }
+            | Instruction::Cls { .. }
+            | Instruction::Rbit { .. }
+            | Instruction::Rev { .. }
+            | Instruction::Rev32 { .. }
+            | Instruction::Rev16 { .. } => true,
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/ir/instructions.rs]: Fixed in `bca621a` (https://github.com/pmatos/s11/pull/109/commits/bca621a).

Changed: `is_encodable_aarch64` now rejects `rd == SP || rn == SP` for all six new bit-manip ops, with the rationale that the up-front filter is consulted by SMT, equivalence, and LLM flows that don't go through the encoder. XZR remains encodable (it's the legitimate AArch64 reg-31 in the Xn slot).
Validation: new `test_is_encodable_bit_manip_rejects_sp` asserts SP rejection on every (rd, rn) position across the six instructions and asserts XZR acceptance; `./ci_check.sh` clean.

Note: the pre-existing MVN/NEG/NEGS arms in the same function have the same `=> true` precedent and would benefit from the same tightening. That cleanup is left for a separate PR so the precedent fix isn't entangled with the new-instruction work.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.x)._

**Comment by pmatos** [comment]
@claude Regarding your third-pass code review (https://github.com/pmatos/s11/pull/109#issuecomment-4457983515) — outcomes below.

| # | Suggestion | Outcome | Reference |
|---|---|---|---|
| 1 | `bv_clz_64` 64-deep ITE chain | DEFER (already tracked) | #112 |
| 2 | Missing positive parser tests | FIX | `bca621a` |
| 3 | Repetitive peer-mutation table | REJECT (already replied last round) | — |
| 4 | Sampling probability (`1/150` per op) | DEFER (already tracked) | #115 |
| 5 | 32-bit W-register forms | DEFER | #121 |
| 6 | `opcode_id` duplication | DEFER (already tracked) | #116 |

---

**FIX:**

- **(#2)** Added six positive parser cases — one per new mnemonic — to `parse_line_covers_all_core_mnemonics` in `src/parser/mod.rs`. The encoder round-trip exercises Capstone disassembly but not the GAS-string parsing path, so this closes a real coverage gap (a future typo in the parse-line dispatch would otherwise only surface via integration tests that don't exist). Validation: `./ci_check.sh` clean.

**DEFER (new):**

- **(#5)** W-register (32-bit) forms → **#121**. Cross-cutting scope expansion — width support doesn't exist anywhere else in the codebase (no W-form for any other instruction), so adding it for the six bit-manip ops alone would be a one-off inconsistency. The proper fix is width-parameterised IR design with a proof-of-concept end-to-end conversion for one instruction first, which is its own piece of work.

**DEFER (already tracked from earlier rounds):**

- **(#1)** `bv_clz_64` ITE depth → #112.
- **(#4)** MCMC sampling distribution → #115. The 1/150 figure you note is exactly what motivated #115; we're keeping the bundled-slot shape until a stochastic-convergence benchmark shows it actually matters.
- **(#6)** `opcode_id` duplication → #116.

**REJECT (already replied to last round):**

- **(#3)** Peer-mutation table verbosity. Same answer as before: the explicit 6-way arms mirror existing MVN/NEG/NEGS, BIC/BICS/ORN/EON, and CSEL clusters in the same file. Collapsing only the new cluster would be inconsistent with the file's existing convention.

---

Also addressed in `bca621a`: a separate codex P2 inline thread on `src/ir/instructions.rs:402` asking that `is_encodable_aarch64` reject SP as rd/rn for the new Xn-class instructions. The encoder already errors on SP via `register_to_dynasm`, but the filter is consulted by paths that bypass the encoder, so SP candidates could previously reach SMT/equivalence/LLM as "encodable". The new arm rejects `rd == SP || rn == SP` while keeping XZR encodable. Regression test added (`test_is_encodable_bit_manip_rejects_sp`).

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.x). If you disagree on any item, please reply and we'll re-evaluate._

Note: this is the third round of bot review on a PR that has been declared "ready to merge" twice. Future feedback raising fresh nits (rather than disagreement with existing triage) will be handled in follow-up issues rather than in-PR commits — at some point new minor suggestions are better resolved post-merge than churning the open PR further.
